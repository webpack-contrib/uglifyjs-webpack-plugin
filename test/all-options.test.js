import { SourceMapSource } from 'webpack-sources';
import UglifyJsPlugin from '../src/index';
import {
  PluginEnvironment,
  cleanErrorStack,
  createCompiler,
  compile,
} from './helpers';

describe('when applied with all options', () => {
  let eventBindings;
  let eventBinding;

  beforeEach(() => {
    const pluginEnvironment = new PluginEnvironment();
    const compilerEnv = pluginEnvironment.getEnvironmentStub();
    compilerEnv.context = '';

    const plugin = new UglifyJsPlugin({
      sourceMap: true,
      extractComments: {
        condition: 'should be extracted',
        filename(file) {
          return file.replace(/(\.\w+)$/, '.license$1');
        },
        banner(licenseFile) {
          return `License information can be found in ${licenseFile}`;
        },
      },
      uglifyOptions: {
        warnings: true,
        mangle: false,
        output: {
          beautify: true,
        },
      },
    });
    plugin.apply(compilerEnv);
    eventBindings = pluginEnvironment.getEventBindings();
  });

  it('matches snapshot', () => {
    const compiler = createCompiler();
    new UglifyJsPlugin({
      sourceMap: true,
      uglifyOptions: {
        mangle: false,
        output: {
          beautify: true,
        },
        warnings: true,
      },
      extractComments: {
        condition: 'should be extracted',
        filename(file) {
          return file.replace(/(\.\w+)$/, '.license$1');
        },
        banner(licenseFile) {
          return `License information can be found in ${licenseFile}`;
        },
      },
    }).apply(compiler);

    return compile(compiler).then((stats) => {
      const errors = stats.compilation.errors.map(cleanErrorStack);
      const warnings = stats.compilation.warnings.map(cleanErrorStack);

      expect(errors).toMatchSnapshot('errors');
      expect(warnings).toMatchSnapshot('warnings');

      for (const file in stats.compilation.assets) {
        if (Object.prototype.hasOwnProperty.call(stats.compilation.assets, file)) {
          expect(stats.compilation.assets[file].source()).toMatchSnapshot(file);
        }
      }
    });
  });

  it('binds one event handler', () => {
    expect(eventBindings.length).toBe(1);
  });

  describe('compilation handler', () => {
    beforeEach(() => {
      [eventBinding] = eventBindings;
    });

    it('binds to compilation event', () => {
      expect(eventBinding.name).toBe('compilation');
    });

    describe('when called', () => {
      let chunkPluginEnvironment;
      let compilationEventBindings;
      let compilationEventBinding;
      let compilation;

      beforeEach(() => {
        chunkPluginEnvironment = new PluginEnvironment();
        compilation = chunkPluginEnvironment.getEnvironmentStub();
        compilation.assets = {
          'test.js': {
            source: () => '/** @preserve Foo Bar */ function foo(longVariableName) { longVariableName = 1; }',
            map: () => {
              return {
                version: 3,
                sources: ['test.js'],
                names: ['foo', 'longVariableName'],
                mappings: 'AAAA,QAASA,KAAIC,kBACTA,iBAAmB',
              };
            },
          },
          'test1.js': {
            source: () => 'invalid javascript',
            map: () => {
              return {
                version: 3,
                sources: ['test1.js'],
                names: [''],
                mappings: 'AAAA',
              };
            },
          },
          'test2.js': {
            source: () => 'function foo(x) { if (x) { return bar(); not_called1(); } }',
            map: () => {
              return {
                version: 3,
                sources: ['test1.js'],
                names: ['foo', 'x', 'bar', 'not_called1'],
                mappings: 'AAAA,QAASA,KAAIC,GACT,GAAIA,EAAG,CACH,MAAOC,MACPC',
              };
            },
          },
          'test3.js': {
            sourceAndMap: () => {
              return {
                source: '/** @preserve Foo Bar */ function foo(longVariableName) { longVariableName = 1; }',
                map: {
                  version: 3,
                  sources: ['test.js'],
                  names: ['foo', 'longVariableName'],
                  mappings: 'AAAA,QAASA,KAAIC,kBACTA,iBAAmB',
                },
              };
            },
          },
          'test4.js': {
            source: () => '/*! this comment should be extracted */ function foo(longVariableName) { /* this will not be extracted */ longVariableName = 1; } // another comment that should be extracted to a separate file\n function foo2(bar) { return bar; }',
            map: () => {
              return {
                version: 3,
                sources: ['test.js'],
                names: ['foo', 'longVariableName'],
                mappings: 'AAAA,QAASA,KAAIC,kBACTA,iBAAmB',
              };
            },
          },
        };
        compilation.errors = [];
        compilation.warnings = [];

        eventBinding.handler(compilation);
        compilationEventBindings = chunkPluginEnvironment.getEventBindings();
      });

      it('binds two event handler', () => {
        expect(compilationEventBindings.length).toBe(2);
      });

      describe('build-module handler', () => {
        beforeEach(() => {
          [compilationEventBinding] = compilationEventBindings;
        });

        it('binds to build-module event', () => {
          expect(compilationEventBinding.name).toBe('build-module');
        });

        it('sets the useSourceMap flag', () => {
          const obj = {};
          compilationEventBinding.handler(obj);
          expect(obj.useSourceMap).toBeTruthy();
        });
      });

      describe('optimize-chunk-assets handler', () => {
        beforeEach(() => {
          [compilationEventBinding] = compilationEventBindings;
        });

        it('binds to optimize-chunk-assets event', () => {
          expect(compilationEventBindings[1].name).toBe('optimize-chunk-assets');
        });

        it('outputs no errors for valid javascript', () => {
          compilationEventBinding.handler([{
            files: ['test.js'],
          }], () => {
            expect(compilation.errors.length).toBe(0);
          });
        });

        it('outputs SourceMapSource for valid javascript', () => {
          compilationEventBinding.handler([{
            files: ['test.js'],
          }], () => {
            expect(compilation.assets['test.js']).toBeInstanceOf(SourceMapSource);
          });
        });

        it('does not output mangled javascript', () => {
          compilationEventBinding.handler([{
            files: ['test.js'],
          }], () => {
            // eslint-disable-next-line no-underscore-dangle
            expect(compilation.assets['test.js']._value).toEqual(expect.stringContaining('longVariableName'));
          });
        });

        it('outputs beautified javascript', () => {
          compilationEventBinding.handler([{
            files: ['test.js'],
          }], () => {
            // eslint-disable-next-line no-underscore-dangle
            expect(compilation.assets['test.js']._value).toEqual(expect.stringContaining('\n'));
          });
        });

        it('does not preserve comments', () => {
          compilationEventBinding.handler([{
            files: ['test.js'],
          }], () => {
            // eslint-disable-next-line no-underscore-dangle
            expect(compilation.assets['test.js']._value).not.toBe(expect.stringContaining('/**'));
          });
        });

        it('outputs parsing errors', () => {
          compilationEventBinding.handler([{
            files: ['test1.js'],
          }], () => {
            expect(compilation.errors.length).toBe(1);
            expect(compilation.errors[0]).toBeInstanceOf(Error);
            expect(compilation.errors[0].message).toEqual(expect.stringContaining('[test1.js:1,0][test1.js:1,8]'));
          });
        });

        it('outputs warnings for unreachable code', () => {
          compilationEventBinding.handler([{
            files: ['test2.js'],
          }], () => {
            expect(compilation.warnings.length).toBe(1);
            expect(compilation.warnings[0]).toBeInstanceOf(Error);
            expect(compilation.warnings[0].message).toEqual(expect.stringContaining('Dropping unreachable code'));
          });
        });

        it('works with sourceAndMap assets as well', () => {
          compilationEventBinding.handler([{
            files: ['test3.js'],
          }], () => {
            expect(compilation.errors.length).toBe(0);
            expect(compilation.assets['test3.js']).toBeInstanceOf(SourceMapSource);
          });
        });

        it('extracts license information to separate file', () => {
          compilationEventBinding.handler([{
            files: ['test4.js'],
          }], () => {
            expect(compilation.errors.length).toBe(0);
            /* eslint-disable no-underscore-dangle */
            expect(compilation.assets['test4.license.js'].source()).toContain('/*! this comment should be extracted */');
            expect(compilation.assets['test4.license.js'].source()).toContain('// another comment that should be extracted to a separate file');
            expect(compilation.assets['test4.license.js'].source()).not.toEqual(expect.stringContaining('/* this will not be extracted */'));
            /* eslint-enable no-underscore-dangle */
          });
        });

        describe('with warningsFilter set', () => {
          describe('and the filter returns true', () => {
            beforeEach(() => {
              const pluginEnvironment = new PluginEnvironment();
              const compilerEnv = pluginEnvironment.getEnvironmentStub();
              compilerEnv.context = '';

              const plugin = new UglifyJsPlugin({
                warningsFilter: () => true,
                sourceMap: true,
                uglifyOptions: {
                  warnings: true,
                  mangle: false,
                  output: {
                    beautify: true,
                  },
                },
              });
              plugin.apply(compilerEnv);
              eventBindings = pluginEnvironment.getEventBindings();

              chunkPluginEnvironment = new PluginEnvironment();
              compilation = chunkPluginEnvironment.getEnvironmentStub();
              compilation.assets = {
                'test2.js': {
                  source: () => 'function foo(x) { if (x) { return bar(); not_called1(); } }',
                  map: () => {
                    return {
                      version: 3,
                      sources: ['test1.js'],
                      names: ['foo', 'x', 'bar', 'not_called1'],
                      mappings: 'AAAA,QAASA,KAAIC,GACT,GAAIA,EAAG,CACH,MAAOC,MACPC',
                    };
                  },
                },
              };
              compilation.errors = [];
              compilation.warnings = [];

              eventBindings[0].handler(compilation);
              compilationEventBindings = chunkPluginEnvironment.getEventBindings();
            });

            it('should get all warnings', () => {
              compilationEventBindings[1].handler([{
                files: ['test2.js'],
              }], () => {
                expect(compilation.warnings.length).toBe(1);
                expect(compilation.warnings[0]).toBeInstanceOf(Error);
                expect(compilation.warnings[0].message).toEqual(expect.stringContaining('Dropping unreachable code'));
              });
            });
          });

          describe('and the filter returns false', () => {
            beforeEach(() => {
              const pluginEnvironment = new PluginEnvironment();
              const compilerEnv = pluginEnvironment.getEnvironmentStub();
              compilerEnv.context = '';

              const plugin = new UglifyJsPlugin({
                warningsFilter: () => false,
                sourceMap: true,
                uglifyOptions: {
                  warnings: true,
                  mangle: false,
                  output: {
                    beautify: true,
                  },
                },
              });
              plugin.apply(compilerEnv);
              eventBindings = pluginEnvironment.getEventBindings();

              chunkPluginEnvironment = new PluginEnvironment();
              compilation = chunkPluginEnvironment.getEnvironmentStub();
              compilation.assets = {
                'test2.js': {
                  source: () => 'function foo(x) { if (x) { return bar(); not_called1(); } }',
                  map: () => {
                    return {
                      version: 3,
                      sources: ['test1.js'],
                      names: ['foo', 'x', 'bar', 'not_called1'],
                      mappings: 'AAAA,QAASA,KAAIC,GACT,GAAIA,EAAG,CACH,MAAOC,MACPC',
                    };
                  },
                },
              };
              compilation.errors = [];
              compilation.warnings = [];

              eventBindings[0].handler(compilation);
              compilationEventBindings = chunkPluginEnvironment.getEventBindings();
            });

            it('should get no warnings', () => {
              compilationEventBindings[1].handler([{
                files: ['test2.js'],
              }], () => {
                expect(compilation.warnings.length).toBe(0);
              });
            });
          });
        });
      });
    });
  });
});
