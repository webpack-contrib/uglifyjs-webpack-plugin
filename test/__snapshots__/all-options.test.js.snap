// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errors 1`] = `Array []`;

exports[`main.6c966be0e6148d71540d.js 1`] = `
"(window.webpackJsonp = window.webpackJsonp || []).push([ [ \\"main\\" ], {
    \\"./test/fixtures/entry.js\\": function(module, exports) {
        eval(\\"// foo\\\\n/* @preserve*/\\\\n// bar\\\\nconst a = 2 + 2;\\\\n\\\\nmodule.exports = function Foo() {\\\\n  const b = 2 + 2;\\\\n  console.log(b + 1 + 2);\\\\n};\\\\n\\\\n\\\\n//////////////////\\\\n// WEBPACK FOOTER\\\\n// ./test/fixtures/entry.js\\\\n// module id = ./test/fixtures/entry.js\\\\n// module chunks = main\\\\n\\\\n//# sourceURL=webpack:///./test/fixtures/entry.js?\\");
    }
}, [ [ \\"./test/fixtures/entry.js\\", \\"manifest\\", \\"main\\" ] ] ]);"
`;

exports[`manifest.ae205c79cb278d0fd374.js 1`] = `
"!function(modules) {
    function webpackJsonpCallback(data) {
        for (var moduleId, chunkId, result, chunkIds = data[0], moreModules = data[1], executeModules = data[2], i = 0, resolves = []; i < chunkIds.length; i++) chunkId = chunkIds[i], 
        installedChunks[chunkId] && resolves.push(installedChunks[chunkId][0]), installedChunks[chunkId] = 0;
        for (moduleId in moreModules) Object.prototype.hasOwnProperty.call(moreModules, moduleId) && (modules[moduleId] = moreModules[moduleId]);
        for (parentJsonpFunction && parentJsonpFunction(data); resolves.length; ) resolves.shift()();
        for (scheduledModules.push.apply(scheduledModules, executeModules || []), i = 0; i < scheduledModules.length; i++) {
            for (var scheduledModule = scheduledModules[i], fullfilled = !0, j = 1; j < scheduledModule.length; j++) {
                var depId = scheduledModule[j];
                0 !== installedChunks[depId] && (fullfilled = !1);
            }
            fullfilled && (scheduledModules.splice(i--, 1), result = __webpack_require__(__webpack_require__.s = scheduledModule[0]));
        }
        return result;
    }
    function __webpack_require__(moduleId) {
        if (installedModules[moduleId]) return installedModules[moduleId].exports;
        var module = installedModules[moduleId] = {
            i: moduleId,
            l: !1,
            exports: {}
        };
        return modules[moduleId].call(module.exports, module, module.exports, __webpack_require__), 
        module.l = !0, module.exports;
    }
    var installedModules = {}, installedChunks = {
        manifest: 0
    }, scheduledModules = [];
    __webpack_require__.e = function(chunkId) {
        var promises = [], installedChunkData = installedChunks[chunkId];
        if (0 !== installedChunkData) if (installedChunkData) promises.push(installedChunkData[2]); else {
            var promise = new Promise(function(resolve, reject) {
                installedChunkData = installedChunks[chunkId] = [ resolve, reject ];
            });
            promises.push(installedChunkData[2] = promise);
            var head = document.getElementsByTagName(\\"head\\")[0], script = document.createElement(\\"script\\");
            script.charset = \\"utf-8\\", script.timeout = 12e4, __webpack_require__.nc && script.setAttribute(\\"nonce\\", __webpack_require__.nc), 
            script.src = __webpack_require__.p + \\"\\" + chunkId + \\".\\" + ({
                main: \\"main\\"
            }[chunkId] || chunkId) + \\".\\" + {
                main: \\"6c966be0e6148d71540d\\"
            }[chunkId] + \\".js\\";
            var timeout = setTimeout(function() {
                onScriptComplete({
                    type: \\"timeout\\",
                    target: script
                });
            }, 12e4);
            script.onerror = script.onload = onScriptComplete;
            function onScriptComplete(event) {
                script.onerror = script.onload = null, clearTimeout(timeout);
                var chunk = installedChunks[chunkId];
                if (0 !== chunk) {
                    if (chunk) {
                        var errorType = event && (\\"load\\" === event.type ? \\"missing\\" : event.type), realSrc = event && event.target && event.target.src, error = new Error(\\"Loading chunk \\" + chunkId + \\" failed.\\\\n(\\" + errorType + \\": \\" + realSrc + \\")\\");
                        error.type = errorType, error.request = realSrc, chunk[1](error);
                    }
                    installedChunks[chunkId] = void 0;
                }
            }
            head.appendChild(script);
        }
        return Promise.all(promises);
    }, __webpack_require__.m = modules, __webpack_require__.c = installedModules, __webpack_require__.d = function(exports, name, getter) {
        __webpack_require__.o(exports, name) || Object.defineProperty(exports, name, {
            configurable: !1,
            enumerable: !0,
            get: getter
        });
    }, __webpack_require__.r = function(exports) {
        Object.defineProperty(exports, \\"__esModule\\", {
            value: !0
        });
    }, __webpack_require__.n = function(module) {
        var getter = module && module.__esModule ? function() {
            return module.default;
        } : function() {
            return module;
        };
        return __webpack_require__.d(getter, \\"a\\", getter), getter;
    }, __webpack_require__.o = function(object, property) {
        return Object.prototype.hasOwnProperty.call(object, property);
    }, __webpack_require__.p = \\"\\", __webpack_require__.oe = function(err) {
        throw console.error(err), err;
    };
    var jsonpArray = window.webpackJsonp = window.webpackJsonp || [], parentJsonpFunction = jsonpArray.push.bind(jsonpArray);
    jsonpArray.push = webpackJsonpCallback, jsonpArray = jsonpArray.slice();
    for (var i = 0; i < jsonpArray.length; i++) webpackJsonpCallback(jsonpArray[i]);
}([]);"
`;

exports[`warnings 1`] = `Array []`;
